public/js/lib/messagePropertyEvaluation.js

```diff
@@ -10,18 +10,45 @@ import * as jmespath from "jmespath";
  * [JMESPath](http://jmespath.org/) syntax is supported.
  */
 export function evaluatePath(path, input) {
+  return evaluate(path, input).value;
+}
+
+/**
+ * Test that the argument is a valid path for evaluatePath.
+ *
+ * @param path {string} The path to validate.
+ */
+export function pathIsValid(path) {
+  return evaluate(path, {}).isValidPath;
+}
+
+function evaluate(path, input) {
   if (!path) {
-    return null;
+    // Falsy paths are valid.
+    // Return null to show an empty card.
+    return { value: null, isValidPath: true };
+  }
+
+  if (path.substring(path.length - 1) === ".") {
+    // jmespath.js has a bug where paths ending in "." aren't rejected by the parser.
+    // https://github.com/jmespath/jmespath.js/issues/36
+    // Return error.
+    return { value: null, isValidPath: false };
   }
 
   try {
-    return jmespath.search(input, path);
+    const value = jmespath.search(input, path);
+    return { value: value, isValidPath: true };
   } catch (error) {
-    if (error.name === "ParserError") {
-      // TODO: The user entered a path that is not a valid JMESPath. This should be detected and prevented earlier in the card settings UI.
-      // Return no match.
-      return null;
+    switch ((error.name || "").toLowerCase()) {
+      case "lexererror":
+      case "parsererror":
```

Comment:
If we don’t know what a `ParserError` or `LexerError` means and we don’t do anything with them, they should not appear in the code at all.

Reply:
Although jmespath.js does not document these in detail, a quick scan of its code reveals that:

- it throws `LexerError` if it fails to tokenize the path, e.g. if the user entered a completely unrecognizable character like `#`
- it throws `ParserError` if it fails to parse the token stream, e.g. if the user entered an incomplete path like `array[`
- it throws `RuntimeError` if its evaluating a bad slice, e.g. `array[0:0:0]`

In all these cases, I think our response is the same: return `null` to the card and report the path is invalid, hence all three fall through to the same `return`.

Leaving the cases out causes these errors to end up in the console. Also I think Vue binding doesn't do so well with an unhandled exception from `pathIsValid`.

Reply:
If the response is the same and we don’t actually handle the cases differently, I suggest to reflect that in the code as well:

```js
return { value: null, isValidPath: false };
throw error;
```

Since the switch case statement doesn’t contain any `break`s, this statement does the same if the error name is one of the three.

Reply:
Sorry, I don't quite follow.

> Since the switch case statement doesn’t contain any breaks, this statement does the same if the error name is one of the three.

Yes. Are you saying that's wrong?

Reply:
In short, switch statements without `break`s are quite weird.

Consider this:

```js
function f(num) {
  switch (num) {
    case 1:
    case 2:
      return 2;
    default:
      return NaN;
  }
}

f(2) // returns 2
f(1) // returns 2
```

Calling the function `f` with either `1` or `2` as an argument will return `2` in both cases! I find that very surprising. When I read this code for `num = 1`, I think it will evaluate the first case statement as true and do nothing since it is empty, but for some reason, the second case statement’s block will be executed as well **without evaluating `case 2`** at all.

To me, that is incredibly unexpected. In our case, I would not expect the code to return that object at all for `ParserError` or `LexerError`. But it does. And that’s very confusing.

Since you want to return the same for all three errors, just do that and drop the switch case statement altogether. It doesn’t provide anything is just very confusing to read.

Reply:
Oh I see your point now. My background includes many curly brace languages where `switch` is used for this specific effect and I did not consider that some contributors might find this behaviour surprising. Banning fall-through cases is not uncommon in style guides - for example Google's C++ and Java style guides ban it in most situations.

I'll change this to an `if`.